pyomo/contrib/pynumero/sparse/mpi_block_matrix.py

```diff
@@ -0,0 +1,1189 @@
+        -------
+        result: BlockVector
+        """
+        block_vector_assert_block_structure(other)
+        assert self.bshape[1] == other.nblocks, 'Dimension mismatch'
+        local_result = BlockVector(self.bshape[0])
+        for row_ndx in range(self.bshape[0]):
+            local_result.set_block(row_ndx, np.zeros(self.get_row_size(row_ndx)))
+        rank = self._mpiw.Get_rank()
+        if rank == 0:
+            block_indices = self._owned_mask
+        else:
+            block_indices = self._unique_owned_mask
+        for row_ndx, col_ndx in zip(*np.nonzero(block_indices)):
+            if self.get_block(row_ndx, col_ndx) is not None:
+                res_blk = local_result.get_block(row_ndx)
+                res_blk += self.get_block(row_ndx, col_ndx) * other.get_block(col_ndx)
+                local_result.set_block(row_ndx, res_blk)
+        flat_local = local_result.flatten()
+        flat_global = np.zeros(flat_local.size)
+        self._mpiw.Allreduce(flat_local, flat_global)
+        global_result = local_result.copy_structure()
+        global_result.copyfrom(flat_global)
+        return global_result
+
+    def __mul__(self, other):
+        assert_block_structure(self)
+
+        if isinstance(other, MPIBlockVector):
+            global_other = other.make_local_copy()
+            result = self._block_vector_multiply(global_other)
+            return result
+        elif isinstance(other, BlockVector):
+            return self._block_vector_multiply(other)
+        elif isinstance(other, np.ndarray):
+            block_other = BlockVector(nblocks=self.bshape[1])
+            for ndx in range(self.bshape[1]):
+                block_other[ndx] = np.zeros(self.get_col_size(ndx), dtype=other.dtype)
+            block_other.copyfrom(other)
+            return self._block_vector_multiply(block_other).flatten()
+        elif np.isscalar(other):
+            result = self.copy_structure()
+            ii, jj = np.nonzero(self._owned_mask)
+            for i, j in zip(ii, jj):
+                if not self._block_matrix.is_empty_block(i, j):
+                    result.set_block(i, j, self.get_block(i, j) * other)
+            return result
+        else:
+            raise NotImplementedError('Operation not supported by MPIBlockMatrix')
+
+    def __rmul__(self, other):
+        assert_block_structure(self)
+        m, n = self.bshape
+        result = self.copy_structure()
+
+        if np.isscalar(other):
+            ii, jj = np.nonzero(self._owned_mask)
+            for i, j in zip(ii, jj):
+                if not self._block_matrix.is_empty_block(i, j):
+                    result.set_block(i, j, self.get_block(i, j) * other)
+            return result
+
+        if isinstance(other, MPIBlockVector):
+            raise NotImplementedError('Vector-Matrix multiply not supported yet')
+        if isinstance(other, BlockVector):
+            raise NotImplementedError('Vector-Matrix multiply not supported yet')
```

Comment:
At some point @carldlaird mentioned we could get some performance benefits if we support xTA. Not sure if thats still in his plans?

Reply:
I added an issue.